tensorpipe/core/pipe.cc

```diff
@@ -85,21 +86,45 @@ void parseDescriptorOfMessage(ReadOperation& op, const Packet& nopPacketIn) {
   }
   for (const auto& nopTensorDescriptor :
        nopMessageDescriptor.tensorDescriptors) {
-    Message::Tensor tensor{
-        .tensor =
-            CpuTensor{
-                .ptr = nullptr,
-                .length = static_cast<size_t>(nopTensorDescriptor.sizeInBytes),
-            },
-        .metadata = nopTensorDescriptor.metadata,
-    };
     ReadOperation::Tensor tensorBeingAllocated;
-    tensorBeingAllocated.length = tensor.tensor.cpu.length;
+    tensorBeingAllocated.length = nopTensorDescriptor.sizeInBytes;
+    tensorBeingAllocated.deviceType = nopTensorDescriptor.deviceType;
     tensorBeingAllocated.channelName = nopTensorDescriptor.channelName;
     // FIXME If the nop object wasn't const we could move the string out...
     tensorBeingAllocated.descriptor = nopTensorDescriptor.channelDescriptor;
-    message.tensors.push_back(std::move(tensor));
     op.tensors.push_back(std::move(tensorBeingAllocated));
+
+    switch (nopTensorDescriptor.deviceType) {
+      case DeviceType::kCpu: {
+        Message::Tensor tensor{
```

Comment:
We could have the cases of the switch just create a tensorpipe::Tensor and then pack it into a tensorpipe::Message::Tensor outside of the switch. This way we wouldn't have to duplicate the metadata stuff, for example.